crates/node/p2p/src/rpc/server.rs

```diff
@@ -0,0 +1,123 @@
+//! RPC Module to serve the P2P API.
+//!
+//! Kona's P2P RPC API is a JSON-RPC API compatible with the [op-node] API.
+//!
+//!
+//! [op-node]: https://github.com/ethereum-optimism/optimism/blob/7a6788836984996747193b91901a824c39032bd8/op-node/p2p/rpc_api.go#L45
+
+use crate::NetRpcRequest;
+use async_trait::async_trait;
+use jsonrpsee::{
+    core::RpcResult,
+    types::{ErrorCode, ErrorObject},
+};
+use kona_rpc::{OpP2PApiServer, PeerDump, PeerInfo, PeerStats};
+use std::net::IpAddr;
+
+/// A type alias for the sender of a [`NetRpcRequest`].
+pub type RpcReqSender = tokio::sync::mpsc::Sender<NetRpcRequest>;
+
+/// NetworkRpc
+///
+/// This is a server implementation of [`OpP2PApiServer`].
+#[derive(Debug)]
+pub struct NetworkRpc {
+    /// The channel to send [`NetRpcRequest`]s.
+    pub sender: RpcReqSender,
+}
+
+impl NetworkRpc {
+    /// Constructs a new [`NetworkRpc`] given a sender channel.
+    pub const fn new(sender: RpcReqSender) -> Self {
+        Self { sender }
+    }
+}
+
+#[async_trait]
+impl OpP2PApiServer for NetworkRpc {
+    async fn opp2p_self(&self) -> RpcResult<PeerInfo> {
+        tracing::trace!("opp2p_self");
+        let (tx, rx) = tokio::sync::oneshot::channel();
+        self.sender
+            .send(NetRpcRequest::PeerInfo(tx))
+            .await
+            .map_err(|_| ErrorObject::from(ErrorCode::InternalError))?;
+
+        rx.await.map_err(|_| ErrorObject::from(ErrorCode::InternalError))
```

Comment:
Good comms pattern.

Not necessary for this PR, but we're going to have _a lot_ of this. Maybe it's worth making a `Request` type that encapsulates this handshake into an easy-to-use abstraction.

Reply:
Good callout, will do